paddle/framework/README.md

```diff
@@ -0,0 +1,87 @@
+# Design Doc: From V2 Python API to Refactorized C++ Core
+
+## Background
+
+We released V2 Python API in February 2017.  This new API would allow users 
+
+1. to write concise programs, and
+1. to write pure Python programs that call PaddlePaddle's C++ core as a dynamic library.
+   
+However, the C++ core code has been old and accumulated messy parts in the recent four years.  Also, the core doesn't implement the key concepts exposed through the V2 Python API.  Therefore, we need to rewrite the C++ core and make sure that it works with the API.
+
+## The V2 API
+
+The design doc of the API is at [here](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/api.md).  For a quick revisit, let's start with a simple program:
+
+```python
+paddle.init(use_gpu=False, trainer_count=1)
+
+x = paddle.layer.data(name='x', type=paddle.data_type.dense_vector(13))
+y_predict = paddle.layer.fc(input=x, size=1, act=paddle.activation.Linear())
+y = paddle.layer.data(name='y', type=paddle.data_type.dense_vector(1))
+cost = paddle.layer.mse(input=y_predict, label=y)
+
+parameters = paddle.parameters.create(cost)
+optimizer = paddle.optimizer.Momentum(momentum=0)
+trainer = paddle.trainer.SGD(cost=cost,
+                             parameters=parameters,
+                             update_equation=optimizer)
+```
+
+## The New C++ Core
+
+The new C++ core implements the following key concepts:
+
+1. Operator, whose inputs and outputs are variables.
+1. Variable, each can hold a value of specified type.
+1. Tensor, a value commonly hold by Variable.
+1. Scope, keep Variables.
+
+## Fill the Gap
+
+We can see the gap between the Python API and the C++ core:
+
+1. the API describes a network layer-by-layer, however, 
+1. the core runs a graph of operators connected by variables in scopes.
+
+We'd expose the core's low-level API to Python via [`pybind11`](https://github.com/pybind/pybind11), and write the high-level API `paddle.layer.xxx` to call the low-level API.
```

Comment:
Or Maybe our high-level API is also an operator in core C++.

As the [`fc_op`](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/fc_op.cc#L23) now in C++ core, we could combine many low-level operators into a high-level operator because `NetOp` is an operator, too.

Just expose `fc_op` to `Python`, and give a proper code generator for `paddle.layer.fc_layer` could fill the gap between layer and operator.

In that code generator, three things should be done.

1. Create Weight Variable. It could be done in Python Code Generator by telling Python which input variable should be weight. i.e., add an `is_weight` to `VarProto` in `OpProto`.
2. Inference Weight Shape. It could be done by C++, if we register a `WeightShapeInferer` method for each layer operator, like `fc_op`.
3. Inference Output Shape. It could be done by `OperatorBase::InferShape`.